blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md

```diff
@@ -0,0 +1,127 @@
+
+Writing tests for a simple logic with a simple input is a breeze. Take for example a test for a simple calculator class. 
+Two numeric inputs and we are done: 
+
+:include-file: com/example/portfolio/SimpleCalculatorTest.java {title: "SimpleCalculatorTest.java"}
+
+However real apps require real domain, and real domain is rarely represented by a couple of numbers.
+Let's take Financial domain as an example. It is full of complex types such as `Transaction` and has a lot of business logic.
+
+One example of such business logic is `ProfitCalculator`. Its job is to calculate profit based on the executed `Transactions`.
+
+After consulting with the business people we are ready to give our implementation a shot.
+
+`Transaction` is a simple data class with a few fields (number of fields is reduced for simplicity).
+
+:include-file: com/example/portfolio/Transaction.java {title: "Transaction.java", readMore: true, readMoreVisibleLines: 20}
+
+Here is our first `ProfitCalculator` implementation, ready to be tested.
+
+:include-file: com/example/portfolio/ProfitCalculator.java {title: "ProfitCalculator.java"}
+
+And here is the first version of our test.
+
+:include-file: com/example/portfolio/ProfitCalculatorWithoutEncapsulationTest.java {title: "ProfitCalculatorTest.java"}
+
+This test is already on a **verbose** side. As the `ProfitCalculator` requirements evolve, the number of setters 
+we need to validate new business logic will grow, as will grow the number of instances that we may need to create. 
+
+Question: But why **verbose** test can be bad?  
+
+```spoiler {title: "tap me to see why verbose test can be a problem"}
+* Verbose test can be bad as it gets harder to spot the test intention due to the setup noise. 
+* Additionally it may discourage you to write separate test scenario and instead you may want to clump test scenarios together and re-use the test data setup.
+* Test data re-use can lead to a random test being broken as you setup data for an unrelated scenario.    
+```
+
+**Verbosity** is not the only potential problem here. The way `Transaction` instance is being constructed for testing 
+purposes can lead to a maintenance burden. 
+
+First iteration of `Transaction` class uses `setters` to set the data. 
+Future iteration may switch to using fluent API instead of `setters`.
+
+:include-java: com/example/portfolio/ProfitCalculatorWithoutEncapsulationFluentTest.java {title: "fluent API", entry: "createTransaction", bodyOnly: true, removeReturn: true}
+
+From the `ProfitCalculator`'s perspective the way `Transaction` instances are created is irrelevant as it doesn't affect the calculation logic.
+But refactoring `Transaction` will break `ProfitCalculatorTest`. 
+
+`Transaction` is going to be used in other tests. As the number of tests that use `Transaction` instances increases you will be more and more reluctant
+to do refactoring. All your tests that create `Transaction` instances are **brittle** tests now.
```

Comment:
Why are they brittle test now? Are they brittle because they hold you back from refactoring? If so, I would specify that.